core/internal/api/credentials.go

```diff
@@ -55,3 +65,237 @@ func (c *apiKeyCredentialProvider) Apply(req *http.Request) error {
+		credentialsFilePath:   settings.GetCredentialsFile(),
+		mu:                    &sync.Mutex{},
+	}, nil
+}
+
+type oauth2CredentialProvider struct {
+	// The URL of the W&B API
+	baseURL string
+
+	// The file path to the JWT
+	identityTokenFilePath string
+
+	// The file path to the access token and its metadata
+	credentialsFilePath string
+
+	// The access token and its metadata
+	token tokenInfo
+
+	mu *sync.Mutex
+}
+
+// ExpiresAt is a custom type representing a time.Time value. It is used to handle
+// expiration times in a specific string format when serializing/deserializing JSON data.
+type ExpiresAt time.Time
+
+const expiresAtLayout = "2006-01-02 15:04:05"
+
+func (e *ExpiresAt) UnmarshalJSON(data []byte) error {
+	var timeString string
+	if err := json.Unmarshal(data, &timeString); err != nil {
+		return err
+	}
+
+	parsedTime, err := time.Parse(expiresAtLayout, timeString)
+	if err != nil {
+		return err
+	}
+
+	*e = ExpiresAt(parsedTime)
+	return nil
+}
+
+func (e ExpiresAt) MarshalJSON() ([]byte, error) {
+	formattedTime := time.Time(e).Format(expiresAtLayout)
+	return json.Marshal(formattedTime)
+}
+
+type tokenInfo struct {
+	// The time at which the token will expire
+	ExpiresAt ExpiresAt `json:"expires_at"`
+
+	// The access token to use for authentication
+	AccessToken string `json:"access_token"`
+}
+
+func (c *tokenInfo) IsTokenExpiring() bool {
+	return time.Until(time.Time(c.ExpiresAt)) <= time.Minute*5
+}
+
+// CredentialsFile is used when serializing/deserializing JSON data from the
+// credentials file
+type CredentialsFile struct {
+	Credentials map[string]tokenInfo `json:"credentials"`
+}
+
+// Apply checks if the access token is expiring, and fetches a new one if so.
+// It then supplies the access token to the request via the Authorization header
+// as a Bearer token
+func (c *oauth2CredentialProvider) Apply(req *http.Request) error {
+	if c.token.IsTokenExpiring() {
+		if err := c.loadCredentials(); err != nil {
+			return err
+		}
+	}
+	req.Header.Set(
+		"Authorization",
+		"Bearer "+c.token.AccessToken,
+	)
+	return nil
+}
+
+// loadCredentials attempts to load an access token from the credentials file.
+// If the credentials file does not exist, it creates it.
+func (c *oauth2CredentialProvider) loadCredentials() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, err := os.Stat(c.credentialsFilePath)
+	if os.IsNotExist(err) {
+		err = c.writeCredentialsFile()
+	}
+	if err != nil {
+		return err
+	}
+
+	return c.loadCredentialsFromFile()
+}
+
+// loadCredentialsFromFile loads the access token from the credentials file. If
+// the access token does not exist for the given base url, or the token is
+// expiring, it fetches a new one from the server, and saves it to the credentials file
+func (c *oauth2CredentialProvider) loadCredentialsFromFile() error {
+	file, err := os.ReadFile(c.credentialsFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read credentials file: %v", err)
+	}
```

Comment:
If we fail to read the file, shouldn't we re-request the token and re-make the file? The user could have deleted the file.

Reply:
We check for the file in the calling function, and write a new one if it doesn't exist.